### ISSUE TITLE
"Custom" resource type

### ISSUE THREAD

**Issue by ethanjdiamond** [SwiftGen/SwiftGen#359]
**Motivation:**
I'm currently using SwiftGen for Strings, Colors and Images. Being able to write your own templates makes it super flexible and really enjoyable to use. One limitation I've hit is that I also want to generate gradients, which would just a list of colors. I'd like to be able to have a file that I define them in like I do colors, and then to have it generate my gradients for me.

**Suggestion:**
It would be nice to have a "Custom" resource. It'd be nice in my yml file to be able to set on or more of these:

```
custom:
    templatePath: [REQUIRED]
    type: "txt"
    paths: [REQUIRED]
    output: [REQUIRED]
```

Where it would parse the keys and values from my file, and let me templatize them through SwiftGen.

I've also noticed you're getting a lot of requests for things like nibs, free-floating images, etc. While they're waiting for you, you could support their needs in the meantime with a custom template:

```
custom:
    templatePath: [REQUIRED]
    type: "file"
    paths: [REQUIRED] (This would be a regex, ex: *.png)
    output: [REQUIRED]

```

In the templates, depending on the type, these would be exposed as "object" objects with either a "key" and "value" or just a "value".

**Comment by djbe** [comment]
We have PRs for `plist` and `json` parsers, which would be useful for what you're describing (1st feature). I'm pretty sure that once #356 is finished (and some related changes), we'll be transferring those PRs from the SwiftGenKit repository to this one, and merging them as soon as possible.

The second feature is also something we've thought about, a loose files parser (that provides only the paths, maybe some metadata (mimetype, size, etc...).

**Comment by djbe** [comment]
Note: we could discuss an improvement to the `colors` parser to handle gradients.

**Comment by AliSoftware** [comment]
Hey, thanks for pitching that suggestion!

I agree that the PRs we already have for YAML/JSON support (or adding a YAML parser, which now that we have the Yams library for the config file already as a dependency will make it trivial, and support JSON too as it's a subset of YAML) will probably already fit your needs/suggestions 👍

Given that this is a long requested feature I think it will probably be one of the first feature we'll add once we finish our big refactoring to monorepo (#356) and made the CI/Rakefile/Danger, etc work again 😉

**Comment by ethanjdiamond** [comment]
Awesome to hear. Thanks!

**Comment by johntmcintosh** [comment]
@AliSoftware @djbe , I currently have a need for a "custom" resource type as well. Similarly in my case, I'm expecting to have a JSON or YAML file which contains some definitions that I would like to be able to parse and run through a template. I can help out with the implementation, but wanted to check in on the status of the existing related PRs (like #288) or other plans before I start working on something. Could you weigh in on what the current plans are, and/or if there's anything in particular I could do to assist?

**Comment by AliSoftware** [comment]
Hi all

I'm quite sorry with the delay on those pending features. End of year has been busy, we've been focused on merging back SwiftGen to a monorepo (and still have some nitpickings to do to finish that) before starting back adding features again, but life came in the way.

Moreover, I'll personally take a well deserved 3 weeks vacancy (8-9 months since my last real vacation!) trekking (so without my laptop), which means that while contributors (especially @djbe) might be around to help and maybe merge some PRs, it's unlikely that there will be a new release until I return mid-february. (That doesn't prevent ideas to be debated and PRs to be reviewed and merged by other core contributors, but just wanted to be transparent about my availability and the release, so that you won't wait in wonder)

Personally for me 2017 has been quite busy and the beginning of 2018 will be too, so reactivity on OSS will be quite slow at least until March (moving out), but my resolution after that will be to catch up with all of that and start being more active on SwiftGen again.

Thanks for understanding and thanks again to all of you for contributing and helping make SwiftGen better!

**Comment by johntmcintosh** [comment]
@AliSoftware, thanks for transparent update, and enjoy the vacation! Fully unplugged time on vacation is always refreshing for me.

If I get a chance I may make a pass at trying to implement the piece that we're in need of right now. I'll probably reference that other PR as I do, and if I'm able to get that into a good state, I'll get a PR up.

**Comment by AliSoftware** [comment]
Thx!

Honestly I think implementing the subcommands to parse arbitrary `yaml` files (which will also support `json` files, as JSON is actually a subset of YAML and any JSON file is a valid YAML file) and `plist` files will solve a lot of features requested by most users, like this PR here, the other PR about custom plist, the recent PR about the `custom` subcommand… all would be solved with one PR to rule them all 😄 so implementing YAML parsing should be the best way to go imho (we already have a pod for that to parse the YAML config file, so you can even use the same lib) to provide both YAML, JSON and env support 3-in-1 :wink:

**Comment by johntmcintosh** [comment]
@AliSoftware , thanks for the recommended approach. I took a pass at what you recommended (PR #379). I have a couple of questions for you guys in the PR description, but overall think that what I have may be a pretty solid start. When you get a chance, I'd love to hear feedback and am happy to update/tweak from there.

By the way - I'm not sure when you're heading out, so feel free to ignore this as needed until you get back.
